### ISSUE TITLE
Develop mechanism for account creators to subscribe to the Capabilities Model Discussion list. 

### ISSUE THREAD

**Issue by dbrunson** [CaRCC/rcd-nexus-portal#201]
As discussed in an email thread around August 29, 2024:

Currently, when someone creates a profile, an email is sent to capsmodel-help and then Dana adds that email address to the capsmodel-discuss list.  (side note: this is occasionally not the email address that they would prefer to be on the list)

The idea is to have an opportunity for everyone who creates a portal account to see an option to subscribe to the discuss list. 

We could just put up a dialog that asks them to opt in, allowing them to specify their preferred email address.  Assuming Tim can figure out an API for this, we could implement with not too much work.

This general approach to asking them about emails has several additional benefits:

- It is explicitly opt-in which seems nicer.
- The popup could appear for every user and not just the creator of a profile.
- We would have a better email for communications that we generate in the portal itself (e.g., requests for access by other team members, etc.).

Lastly - we should also ask Tim about tying this into Wild Apricot for the abilities it also gives us.

**Comment by dbrunson** [comment]
@MiddelkoopT  - we welcome your thoughts on this :)

**Comment by middelkoopt** [comment]
All very doable.  Just create a REST API endpoint, let me know where the docs are and an API key and I'll make it happen.  Alternatively we could use a web-hook with a JSON payload (with a API key).  I don't see this requiring much work to integrate into Wild Apricot and Google Groups

Here is an off-the-cuff sketch of an API
```
GET /v1/accounts?synchronized=false
```

This returns
```json
[ {
  "group": cappsmodel-help@carcc.org
  "name": <name>,
  "email": <email>,
  "google": [ <email>, <email>],
  "synchronized": false
}, {
...
]
```

```
PATCH /v1/account/<email>
BODY:
```
```json
{ "group": "cappsmodel-help@carcc.org", "email": <email>, "synchronized": true}
```

returns
true
on success, false otherwise

Wild apricot would use an internal field to populate from this information.  The synchronization would happen with the main WA sync job (every 10 min).

**Comment by plschmitz** [comment]
Ideally, we would just issue the request with no ID bindings (e.g., email and list name), rather than the more proper RESTful interface with an account identifier. Is that possible?

 

My wish would be for a quiet fail if they are already subscribed, and that the request be async if it cannot return very quickly.  We do not need to give the user a response (other than some generic “Your request to be added to the group has been sent”), although a success/failure response would be nice if the call returns quickly. 

 

Does that seem reasonable, or not?

 

Thanks - Patrick

**Comment by middelkoopt** [comment]
Not sure what you mean "Ideally, we would just issue the request with no ID bindings (e.g., email and list name), rather than the more proper RESTful interface with an account identifier. Is that possible?"

The API I describing was on the caps model side.

I could also provide an endpoint.  It would be pretty quick.  If you wanted just the google group, I have code for that as well. The google group API is pretty simple.

Are you thinking something like this?
```
PUT /v1/subscribe?email=<email>&group=<list-email>
```
returns "true" or "false" ?

The issue with WA is that I need their primary email id to set things.  I can lookup other accounts, but I don't think I have a library for that yet but it should not be that difficult.

Also, I think you could have Wild Apricot be a login provider as well.
something like https://www.drupal.org/docs/contributed-modules/drupal-oauth-oidc-login/wild-apricot-sso-login
I think there might be other integrations as well.

**Comment by plschmitz** [comment]
Yes, I was thinking something like the PUT you describe. I don’t want to take the time to develop a UI to negotiate with users what their WA primary account email is. 

 

Is it a problem (for your model of groups management and WA) to hit the google groups directly as you propose?

 

Let’s talk separately about WA as an IdP.

 

Thanks - Patrick

**Comment by middelkoopt** [comment]
It is easy to hit google groups via my API, but you could also just do this directly with the Google Groups API
https://developers.google.com/admin-sdk/directory/reference/rest/v1/members/insert

I could assist in that development and provide the properly scoped API keys/etc.

**Comment by plschmitz** [comment]
Here's the call: POST https://admin.googleapis.com/admin/directory/v1/groups/NNNNN/members
where "NNNNN" is the group key, which [according to docs](https://developers.google.com/admin-sdk/directory/v1/guides/manage-groups) can be the email (capsmodel-discuss@carcc.org) with this minimal payload:
```
{
       "email": "liz@example.com",
        "role": "MEMBER"
}
```
Another [useful code snippet](https://stackoverflow.com/questions/11663945/calling-a-rest-api-from-django-view)
Look into [auth](https://developers.google.com/docs/api/quickstart/python#configure_the_oauth_consent_screen
), [python api](https://cloud.google.com/python/docs/reference), etc.

**Comment by middelkoopt** [comment]
More examples from the CaRCC membership automation code: 
Group code https://github.com/CaRCC/CaRCC-Automation/blob/main/Membership/CaRCC_Membership/Groups.js
Wild Apricot code https://github.com/CaRCC/CaRCC-Automation/blob/main/Membership/CaRCC_Membership/Apricot.js
